Split load/store instructions of LLVM struct types

shaobo-he opened the pull request:
This pull request intends to model load/store instructions of LLVM struct types. Previously, SMACK translates a load instruction to a variable with an aggregate type into a load function returning `ref` type and  does not allow store instructions with aggregate store values. With this pull request, a load instruction of a struct type is translated into a series of l`oad` instructions and `insertvalue` instructions while a store instruction is translated into a series of `getelementptr` instructions (optional when the value operand is a constant) and `store` instructions. For example,

```llvm
 store { i64, i64 } { i64 ptrtoint (i8* (%"class.Memory::WriteBarrierPtr.8"*, i64)* @_ZN6Memory13HeapAllocator5AllocEm to i64), i64 0 }, { i64, i64 }* %10, align 8, !dbg !49654
```
==>
```llvm
  %20 = getelementptr { i64, i64 }, { i64, i64 }* %5, i32 0, i32 1, !dbg !42958
  store i64 0, i64* %20, !dbg !42958                                                                  
  %21 = getelementptr { i64, i64 }, { i64, i64 }* %5, i32 0, i32 0, !dbg !42958
  store i64 ptrtoint (i8* (%"class.Memory::WriteBarrierPtr.8"*, i64)* @_ZN6Memory13HeapAllocator5AllocEm to i64), i64* %21, !dbg !42958 
```
I added an option `--split-structs` such that the pass can be invoked on demand.

michael-emmi commented:
This PR needs a simple test which only succeeds with the `--split-structs` flag enabled. I came up with the following test which I thought would do the trick, but it doesn't:
````
#include <smack.h>

typedef struct {
  int x;
  int y;
  int z;
} S;

S foo() {
  S s;
  s.x = __VERIFIER_nondet_int();
  s.y = 2;
  s.z = __VERIFIER_nondet_int();
  return s;
}

int main() {
  S s = foo();
  assert(s.y == 2);
  return 0;
}
````
It seems like structs are indeed split in the translation with `--split-structs`, however Smack still reports a failed assertion.

shaobo-he commented:
Hi @michael-emmi,

Thanks for taking a look at this pull request. Yes, I agree that we need a regression test. However, it seems to me that it's much easier to come up with a program in bitcode that exercises this PR. The example in your comment seems to require bit-precision since the first two fileds, `x,y` in `S` are coerced into an `i64`. How about using the following slightly modified version of the example?
```C
#include <smack.h>

typedef struct {
  int x;
  int y;
  int z;
} S;

S foo() {
  S s;
  s.x = 1;
  s.y = 2;
  s.z = 3;
  return s;
}

int main() {
  S s = foo();
  assert(s.z == 3);
  return 0;
}
```

On my machine, the assertions fails w.o. `--split-structs` while does not otherwise.

michael-emmi commented:
Sure, add that example then. Remember to add the line
````
// @flag --split-structs
````
to the regression.

michael-emmi commented on include/smack/SplitStructLoadStore.h:
```diff
@@ -0,0 +1,31 @@
+
+namespace smack {
+
+class SplitStructLoadStore : public llvm::ModulePass {
```
According to the [LLVM docs](http://llvm.org/docs/WritingAnLLVMPass.html#pass-classes-and-requirements) it seems like this should be a `BasicBlockPass` and not a `ModulePass`:

> When choosing a superclass for your Pass, you should choose the most specific class possible, while still being able to meet the requirements listed. This gives the LLVM Pass Infrastructure information necessary to optimize how passes are run, so that the resultant compiler isn’t unnecessarily slow.

michael-emmi commented on include/smack/SplitStructLoadStore.h:
```diff
@@ -0,0 +1,31 @@
+  SplitStructLoadStore() : llvm::ModulePass(ID) {}
+  virtual bool runOnModule(llvm::Module& M);
+private:
+  const llvm::DataLayout* DL;
```
This variable is initialized and never used.

michael-emmi commented on lib/smack/SplitStructLoadStore.cpp:
```diff
@@ -0,0 +1,151 @@
+
+bool SplitStructLoadStore::runOnModule(Module& M)
+{
+  DL = &M.getDataLayout();
```
This variable (`DL`) is initialized and never used.

michael-emmi commented on include/smack/SplitStructLoadStore.h:
```diff
@@ -0,0 +1,31 @@
+  virtual bool runOnModule(llvm::Module& M);
+private:
+  const llvm::DataLayout* DL;
+  std::vector<llvm::Instruction*> toRemove;
```
Couldn't this be a local variable — e.g., in what is currently the `runInModule` function — rather than an instance variable?

michael-emmi commented on share/smack/top.py:
```diff
@@ -175,6 +175,9 @@ def arguments():
   translate_group.add_argument('--float', action="store_true", default=False,
     help='enable bit-precise floating-point functions')
 
+  translate_group.add_argument('--split-structs', action='store_true', default=False,
```
Do we really want this to be optional? Are there any bad things that can happen by splitting structs into individual accesses? Clearly not doing it is less precise.

shaobo-he replied:
Let's make it optional for now since we haven't tested this flag extensively on SVCOMP benchmarks. If the results look fine, we can remove this flag.

michael-emmi commented on lib/smack/SplitStructLoadStore.cpp:
```diff
@@ -0,0 +1,151 @@
+    std::vector<std::pair<Value*, unsigned> > idx;
+    IRBuilder<> irb(li);
+    li->replaceAllUsesWith(buildStructs(&irb, li->getPointerOperand(), st, nullptr, idx));
+    toRemove.push_back(li);
```
This could just as easily be done at the call site of `splitStructLoad`, making `toRemove` a local variable.

michael-emmi commented on lib/smack/SplitStructLoadStore.cpp:
```diff
@@ -0,0 +1,151 @@
+    std::vector<std::pair<Value*, unsigned> > idx;
+    IRBuilder<> irb(si);
+    copyStructs(&irb, ptr, st, val, idx);
+    toRemove.push_back(si);
```
This could just as easily be done at the call site of `splitStructStore`, making `toRemove` a local variable.

shaobo-he commented:
@michael-emmi Thanks for your suggestions. I've refactored the code according to your comments and am waiting for the result of Travis CI.

zvonimir commented:
Hi Shaobo. Could run please run our usual SVCOMP extensive benchmarking steps to make sure this does not break anything? Please report back once that is done. Thanks!

zvonimir commented on lib/smack/SplitStructLoadStore.cpp:
```diff
@@ -0,0 +1,148 @@
+
+using namespace llvm;
+
+std::vector<Value*> getFirsts(std::vector<std::pair<Value*, unsigned> > lst)
```
You have a coding style mistake here - we put open curly braces after function definition, meaning after `)`. Take a look at other files for examples and then please fix this.

zvonimir replied:
You probably don't need this space after `unsigned>`.

zvonimir commented on include/smack/SplitStructLoadStore.h:
```diff
@@ -0,0 +1,29 @@
+
+private:
+  void splitStructLoad(llvm::LoadInst* li);
+  llvm::Value* buildStructs(llvm::IRBuilder<> *irb, llvm::Value* ptr, llvm::Type* ct, llvm::Value* val, std::vector<std::pair<llvm::Value*, unsigned> > idxs);
```
Please break these long lines somewhere around 80-100 characters.

zvonimir commented on lib/smack/SplitStructLoadStore.cpp:
```diff
@@ -0,0 +1,148 @@
+std::vector<unsigned> getSeconds(std::vector<std::pair<Value*, unsigned> > lst)
+{
+    std::vector<unsigned> ret;
+    for (auto p = lst.begin()+1; p != lst.end(); ++p)
```
Wait, what happens here if `lst.begin == lst.end`?? Meaning, when you have only one element in this list. Is that situation possible?
If not, then it would be good to add an assertion here to make sure.

zvonimir commented on lib/smack/SplitStructLoadStore.cpp:
```diff
@@ -0,0 +1,148 @@
+    if (LoadInst* li = dyn_cast<LoadInst>(&I)) {
+      bool isReturnValue = false;
+      for (auto u = li->user_begin(); u != li->user_end(); ++u)
+      {
```
Again, fix where you open curly braces.

zvonimir commented on lib/smack/SplitStructLoadStore.cpp:
```diff
@@ -0,0 +1,148 @@
+void SplitStructLoadStore::splitStructLoad(LoadInst* li)
+{
+  if (StructType* st = dyn_cast<StructType>(li->getType())) {
+    std::vector<std::pair<Value*, unsigned> > idx;
```
No space needed after `unsigned>`.

zvonimir commented on lib/smack/SplitStructLoadStore.cpp:
```diff
@@ -0,0 +1,148 @@
+  }
+}
+
+Value* SplitStructLoadStore::buildStructs(IRBuilder<> *irb, Value* ptr, Type* ct, Value* val, std::vector<std::pair<Value*, unsigned> > idxs)
```
Split this long line.

zvonimir commented on lib/smack/SplitStructLoadStore.cpp:
```diff
@@ -0,0 +1,148 @@
+        }
+      }
+      if (!isReturnValue && li->getType()->isAggregateType()) {
+        splitStructLoad(li);
```
Hmmm, in `splitStructLoad` you do not always perform this splitting since this function start with an if condition. And yet here you always add this instruction into `toRemove`. Is this the right thing to do even when line 58 evaluates to false?

shaobo-he replied:
Thanks for the good catch. I just realized this file should be named as `SplitAggregateLoadStore` since it's supposed to handle both `struct` and `array` types. The reason I made this mistake is that it seems really hard to find test cases for `array` types. All the examples that I've seen is about `struct` types. I just removed the conditions in `splitStructLoad`, which means this pass handles both types while is not named accordingly.

zvonimir replied:
Wait, why don't you change the file and class name then? Would it not make sense to do that and it is easy to do it?

zvonimir commented on lib/smack/SplitStructLoadStore.cpp:
```diff
@@ -0,0 +1,148 @@
+
+  if (ct->isIntegerTy() || ct->isPointerTy() || ct->isFloatingPointTy())
+    return irb->CreateInsertValue(val,
+             irb->CreateLoad(irb->CreateGEP(ptr, ArrayRef<Value*>(getFirsts(idxs)))),
```
This is not how we align (in general) when we split into multiple lines. See other files for examples.

shaobo-he replied:
I think the way that this line is split makes sense to me because each argument of the top level function call is put into separate lines. It should help readability.

zvonimir commented on lib/smack/SplitStructLoadStore.cpp:
```diff
@@ -0,0 +1,148 @@
+             irb->CreateLoad(irb->CreateGEP(ptr, ArrayRef<Value*>(getFirsts(idxs)))),
+               ArrayRef<unsigned>(getSeconds(idxs)));
+  else if (ArrayType* AT = dyn_cast<ArrayType>(ct)) {
+    for (unsigned i = AT->getNumElements(); i-- > 0; ) {
```
This for loop is a bit weird :). Why is `i--` not in the end as usual?

shaobo-he replied:
I think the for loop is equivalent to ```for (unsigned i = AT->getNumElements() - 1; i >= 0; --i)```.

zvonimir replied:
I understand they are semantically equivalent, but the way you implemented this loop is very unusual, and it is better to switch to a more standard way. In addition, LLVM coding guidelines prefer preincrement/decrement to postincrement/decrement:
http://llvm.org/docs/CodingStandards.html#prefer-preincrement
So please change this loop to what you suggested in your message above.

zvonimir commented on lib/smack/SplitStructLoadStore.cpp:
```diff
@@ -0,0 +1,148 @@
+void SplitStructLoadStore::splitStructStore(StoreInst* si, Value* ptr, Value* val)
+{
+  if (StructType* st = dyn_cast<StructType>(val->getType())) {
+    std::vector<std::pair<Value*, unsigned> > idx;
```
Space after `unsigned` is not needed.

zvonimir commented on lib/smack/SplitStructLoadStore.cpp:
```diff
@@ -0,0 +1,148 @@
+      cv = buildStructs(irb, ptr, AT->getElementType(), cv, lidxs);
+    }
+  }
+  else if (StructType* ST = dyn_cast<StructType>(ct)) {
```
else if should be on the previous line after curly brace.

zvonimir commented on lib/smack/SplitStructLoadStore.cpp:
```diff
@@ -0,0 +1,148 @@
+      cv = buildStructs(irb, ptr, ST->getElementType(i), cv, lidxs);
+    }
+  }
+
```
What happens if neither if branch nor else if branch above evaluate to true?
Good programming practice is to create an else branch and put llvm_unreachable in there as a sanity check. Grep for llvm_unreachable to see how to use it.

zvonimir commented on test/basic/split_structs.c:
```diff
@@ -0,0 +1,27 @@
+#include "smack.h"
+
+//There exists a load/store instruction of type {i64, i32}
```
What does this comment mean? Please refine it. Also, put space after `//`.

zvonimir commented on test/basic/split_structs_fail.c:
```diff
@@ -0,0 +1,27 @@
+#include "smack.h"
+
+//There exists a load/store instruction of type {i64, i32}
```
Look for my comment above.

zvonimir commented on include/smack/SplitStructLoadStore.h:
```diff
@@ -0,0 +1,30 @@
+  llvm::Value* buildStructs(llvm::IRBuilder<> *irb, llvm::Value* ptr, llvm::Type* ct,
+                              llvm::Value* val, std::vector<std::pair<llvm::Value*, unsigned>> idxs);
+  void splitStructStore(llvm::StoreInst* si, llvm::Value* ptr, llvm::Value* val);
+  void copyStructs(llvm::IRBuilder<> *irb, llvm::Value* ptr, llvm::Type* ct, llvm::Value* val, std::vector<std::pair<llvm::Value*, unsigned>> idxs);
```
Split this long line. See LLVM coding guidelines.

zvonimir commented on lib/smack/SplitStructLoadStore.cpp:
```diff
@@ -0,0 +1,137 @@
+
+std::vector<Value*> getFirsts(std::vector<std::pair<Value*, unsigned>> lst) {
+    std::vector<Value*> ret;
+    for (auto p = lst.begin(); p != lst.end(); ++p)
```
Do you need `begin` and `end` here? Can's you just write `for (auto p : lst)`?

zvonimir commented on lib/smack/SplitStructLoadStore.cpp:
```diff
@@ -0,0 +1,137 @@
+  std::vector<Instruction*> toRemove;
+  for(Instruction& I : BB) {
+    if (LoadInst* li = dyn_cast<LoadInst>(&I)) {
+      bool isReturnValue = false;
```
Rewrite it like this:
http://llvm.org/docs/CodingStandards.html#turn-predicate-loops-into-predicate-functions

zvonimir commented on lib/smack/SplitStructLoadStore.cpp:
```diff
@@ -0,0 +1,137 @@
+
+using namespace llvm;
+
+std::vector<Value*> getFirsts(std::vector<std::pair<Value*, unsigned>> lst) {
```
Wait, shouldn't there be a more succinct way of doing this?
There must be a library function or something like that that helps you to extract first elements of this vector of pairs, no?
I googled it and this is one potentially nice hit: https://stackoverflow.com/questions/9094132/c-stdtransform-vector-of-pairs-first-to-new-vector

shaobo-he replied:
I think this function is fine because C++ does not offer simple built-in functions to do mapping from one list to another. The methods discussed in the link also require at least a function to be written.

zvonimir commented on lib/smack/SplitStructLoadStore.cpp:
```diff
@@ -0,0 +1,137 @@
+
+std::vector<unsigned> getSeconds(std::vector<std::pair<Value*, unsigned>> lst) {
+    std::vector<unsigned> ret;
+    for (auto p = lst.begin()+1; p != lst.end(); ++p)
```
What happens when `lst.begin() == lst.end()`?? Is that possible?

shaobo-he replied:
I don't think it's possible as long as we make sure that this pass always works on aggregate types, which has been done.

zvonimir commented on lib/smack/SplitStructLoadStore.cpp:
```diff
@@ -0,0 +1,137 @@
+      cv = buildStructs(irb, ptr, ST->getElementType(i), cv, lidxs);
+    }
+  } else
+    llvm_unreachable("Unsupported types");
```
This string should probably be `Unsupported type.`

zvonimir commented on lib/smack/SplitStructLoadStore.cpp:
```diff
@@ -0,0 +1,137 @@
+      copyStructs(irb, ptr, ST->getElementType(i), A, lidxs);
+    }
+  } else
+    llvm_unreachable("Unsupported types");
```
This string should probably be `Unsupported type`.

shaobo-he commented:
@zvonimir Thanks for the suggestions. I've addressed most of issues that we agree on. Please take a look at it. If it's fine, then I will rename this file to reflect that fact that this PR should work on aggregate types.

zvonimir commented on lib/smack/SplitStructLoadStore.cpp:
```diff
@@ -0,0 +1,140 @@
+}
+
+static bool isUsedByReturnInst(Value* v)
+{
```
Wrong place for this curly brace.
